### PR TITLE
Eip1234

### PR DESCRIPTION
**Pull request by holiman**
Replaces https://github.com/ethereum/go-ethereum/pull/17657 (but maintains the commit history from @eosclassicteam ). I wasn't sure you wanted me to push directly into your branch -- this PR contains a bit less copy-paste.

### CODE REVIEW

File: consensus/ethash/consensus.go [hunk 1]
```diff
@@ -38,10 +38,21 @@ import (
...
-	allowedFutureBlockTime = 15 * time.Second  // Max time from current time allowed for blocks, before they're considered future blocks
+	FrontierBlockReward       = big.NewInt(5e+18) // Block reward in wei for successfully mining a block
+	ByzantiumBlockReward      = big.NewInt(3e+18) // Block reward in wei for successfully mining a block upward from Byzantium
+	ConstantinopleBlockReward = big.NewInt(2e+18) // Block reward in wei for successfully mining a block upward from Constantinople
+	maxUncles                 = 2                 // Maximum number of uncles allowed in a single block
+	allowedFutureBlockTime    = 15 * time.Second  // Max time from current time allowed for blocks, before they're considered future blocks
+
+	// calcDifficultyConstantinople is the difficulty adjustment algorithm for Constantinople.
+	// It returns the difficulty that a new block should have when created at time given the
+	// parent block's time and difficulty. The calculation uses the Byzantium rules, but with bomb offset 5M.
```
Comment (karalabe) [inline consensus/ethash/consensus.go]: Nitpick, please move `bomb offset 5M` to a new line.

I'd also perhaps use 5M and 3M instead and do the decrement internally. It's not really that expensive imho but the code would look a lot cleaner opposed to guessing why the `-1` here opposed to the EIPs.

Also please add a link to the two iceage postponing EIPs here for reference. Some future maintainer will thank us.

File: consensus/ethash/consensus.go [hunk 2]
```diff
@@ -335,6 +338,10 @@ var (
 // the difficulty is calculated with Byzantium rules, which differs from Homestead in
 // how uncles affect the calculation
 func makeDifficultyCalculator(bombDelay *big.Int) func(time uint64, parent *types.Header) *big.Int {
+	// Note, the calculations below looks at the parent number, which is 1 below
+	// the block number. Thus we remove one from the delay given
+	bombDelayFromParent := new(big.Int)
+	bombDelayFromParent = bombDelay.Sub(bombDelay, big1)
```
Comment (karalabe) [inline consensus/ethash/consensus.go]: bombDelayFromParent := new(big.Int).Sub(bombDelay, big1)

Comment (holiman) [reply consensus/ethash/consensus.go]: gotcha, done

**Comment by karalabe** [comment]
@eosclassicteam While in general that's not a bad idea, within Geth we kind of always focused on the Ethereum use case and tried to avoid making everything too abstract/flexible. There have been multiple consensus issues in Parity caused by EIPs getting accidentally activated due to their flexible architecture, as well as configuring a Parity node is a nightmare that can lately only be done via custom tooling. Our code mostly assumes that Ethereum hard forks come in progression one after the other and this keeps everything a lot simpler. I of course understand that this puts a bit more burden on teams creating Ethereum forks, but at the end of the day why should the go-ethereum maintainers pay the price of running derivatives based on our code? :)

**Comment by karalabe** [comment]
For dev/private networks we already have the Clique engine, so PoW difficulty/rewards are a lot less relevant.

**Pull request merged**